rust/parquet_derive/src/parquet_field.rs

```diff
@@ -0,0 +1,790 @@
+                        Type::TypePath(_) => second_type,
+                        Type::Option(ref fourth_type)
+                        | Type::Vec(ref fourth_type)
+                        | Type::Array(ref fourth_type)
+                        | Type::Reference(_, ref fourth_type) => match **fourth_type {
+                            Type::TypePath(_) => third_type,
+                            _ => unimplemented!("sorry, I don't descend this far"),
+                        },
+                    },
+                },
+            },
+        }
+    }
+
+    pub fn last_part(&self) -> String {
+        let inner_type = self.inner_type();
+        let inner_type_str = (quote! { #inner_type }).to_string();
+
+        inner_type_str.split("::").last().unwrap().trim().to_string()
+    }
+
+    pub fn physical_type(&self) -> parquet::basic::Type {
+        use parquet::basic::Type as BasicType;
+
+        let last_part = self.last_part();
+        let leaf_type = self.leaf_type();
+
+        match leaf_type {
+            Type::Array(ref first_type) => {
+                if let Type::TypePath(_) = **first_type {
+                    if last_part == "u8" {
+                        return BasicType::FIXED_LEN_BYTE_ARRAY;
+                    }
+                }
+            }
+            Type::Vec(ref first_type) => {
+                if let Type::TypePath(_) = **first_type {
+                    if last_part == "u8" {
+                        return BasicType::BYTE_ARRAY;
+                    }
+                }
+            }
+            _ => (),
+        }
+
+        match last_part.trim() {
+            "bool" => BasicType::BOOLEAN,
+            "u8" | "u16" | "u32" => BasicType::INT32,
+            "i8" | "i16" | "i32" => BasicType::INT32,
+            "u64" | "i64" | "usize" | "NaiveDateTime" => BasicType::INT64,
+            "f32" => BasicType::FLOAT,
+            "f64" => BasicType::DOUBLE,
+            "String" | "str" => BasicType::BYTE_ARRAY,
+            f @ _ => unimplemented!("sorry, don't handle {} yet!", f),
+        }
+    }
+
+    pub fn from_type_path(f: &syn::Field, p: &syn::TypePath) -> Self {
+        let segments_iter = p.path.segments.iter();
+        let segments: Vec<syn::PathSegment> = segments_iter.map(|x| x.clone()).collect();
+        let last_segment = segments.last().unwrap();
+
+        let is_vec =
+            last_segment.ident == syn::Ident::new("Vec", proc_macro2::Span::call_site());
+        let is_option = last_segment.ident
+            == syn::Ident::new("Option", proc_macro2::Span::call_site());
+
+        if is_vec || is_option {
+            let generic_type = match &last_segment.arguments {
+                syn::PathArguments::AngleBracketed(angle_args) => {
+                    let mut gen_args_iter = angle_args.args.iter();
+                    let first_arg = gen_args_iter.next().unwrap();
+                    assert!(gen_args_iter.next().is_none());
```

Comment:
Can these be simplified to?
```rust
let first_arg = &angle_args.args[0];
```

Reply:
I want to make sure there is only one generic argument to the type, it's just an assertion for preconditions so the code doesn't mysteriously break further in the code generation. We don't support things like `Map<&str, bool>`, for example. Perhaps it should be a different kind of error? This is the only `assert!` in the code so I think I should go to the more standard `unimplemented!()`

Reply:
In that case you can use this:
```
assert!(angle_args.args.len() == 1);
let first_arg = &angle_args.args[0];
```
It's better than using iterator.

Reply:
Good point.